File: core/src/com/ore/infinium/systems/PlayerSystem.java
package com.ore.infinium.systems;

import com.artemis.Aspect;
import com.artemis.ComponentMapper;
import com.artemis.annotations.Wire;
import com.artemis.systems.IteratingSystem;
import com.badlogic.gdx.math.Vector2;
import com.ore.infinium.LoadedViewport;
import com.ore.infinium.OreTimer;
import com.ore.infinium.OreWorld;
import com.ore.infinium.components.*;

/**
 * ***************************************************************************
 * Copyright (C) 2015 by Shaun Reich <sreich02@gmail.com>                    *
 * *
 * This program is free software; you can redistribute it and/or            *
 * modify it under the terms of the GNU General Public License as           *
 * published by the Free Software Foundation; either version 2 of           *
 * the License, or (at your option) any later version.                      *
 * *
 * This program is distributed in the hope that it will be useful,          *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 * GNU General Public License for more details.                             *
 * *
 * You should have received a copy of the GNU General Public License        *
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.    *
 * ***************************************************************************
 */
@Wire(failOnNull = false)
public class PlayerSystem extends IteratingSystem {
    private OreWorld m_world;

    private ComponentMapper<PlayerComponent> playerMapper;
    private ComponentMapper<SpriteComponent> spriteMapper;
    private ComponentMapper<ControllableComponent> controlMapper;
    private ComponentMapper<ItemComponent> itemMapper;
    private ComponentMapper<VelocityComponent> velocityMapper;
    private ComponentMapper<JumpComponent> jumpMapper;

    private NetworkServerSystem m_networkServerSystem;

    private OreTimer chunkTimer = new OreTimer();

    public PlayerSystem(OreWorld world) {
        super(Aspect.one(PlayerComponent.class));

        m_world = world;

    }

    @Override
    protected void inserted(int entityId) {
        super.inserted(entityId);

        //client does nothing as of yet, with this
        if (m_world.isClient()) {
            return;
        }

        //initial spawn, send region
        calculateLoadedViewport(entityId);
        sendPlayerBlockRegion(entityId);
    }

    @Override
    protected void removed(int entityId) {
        super.removed(entityId);

    }

    @Override
    protected void process(int entityId) {
        if (m_world.isClient()) {
            return;
        }

        //clients, for now, do their own collision stuff. mostly.
        //FIXME: clients should simulate their own player's collision with everything and tell the server its
        // position so it can broadcast.
        // but nothing else.
        //server will simulate everything else(except players), and broadcast positions

        //should never ever, ever happen.
        assert spriteMapper.has(entityId) && playerMapper.has(entityId);

        SpriteComponent spriteComponent = spriteMapper.getSafe(entityId);
        PlayerComponent playerComponent = playerMapper.get(entityId);

        com.badlogic.gdx.math.Rectangle viewportRect = playerComponent.loadedViewport.rect;
        float x = spriteComponent.sprite.getX();
        float y = spriteComponent.sprite.getY();

        //fixme fixme, we'll fix this when we get to chunking and come up with a proper solution
        if (chunkTimer.milliseconds() > 600) {
            calculateLoadedViewport(entityId);
            chunkTimer.reset();
        }
    }

    private void calculateLoadedViewport(int playerEntity) {
        PlayerComponent playerComponent = playerMapper.get(playerEntity);
        SpriteComponent spriteComponent = spriteMapper.get(playerEntity);

        LoadedViewport loadedViewport = playerComponent.loadedViewport;

        Vector2 center = new Vector2(spriteComponent.sprite.getX(), spriteComponent.sprite.getY());
        loadedViewport.centerOn(center);

        m_networkServerSystem.sendPlayerLoadedViewportMoved(playerEntity);
        sendPlayerBlockRegion(playerEntity);
    }

    private void sendPlayerBlockRegion(int playerEntity) {
        PlayerComponent playerComponent = playerMapper.get(playerEntity);
        LoadedViewport loadedViewport = playerComponent.loadedViewport;

        LoadedViewport.PlayerViewportBlockRegion region = loadedViewport.blockRegionInViewport();

        m_networkServerSystem.sendPlayerBlockRegion(playerEntity, region.x, region.y, region.width, region.height);
    }

}

File: core/src/com/ore/infinium/systems/PlayerSystem.kt
/**
MIT License

Copyright (c) 2016 Shaun Reich <sreich02@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */

package com.ore.infinium.systems

import com.artemis.Aspect
import com.artemis.annotations.Wire
import com.artemis.systems.IteratingSystem
import com.badlogic.gdx.math.Vector2
import com.ore.infinium.OreTimer
import com.ore.infinium.OreWorld
import com.ore.infinium.components.PlayerComponent
import com.ore.infinium.components.SpriteComponent
import com.ore.infinium.systems.server.ServerNetworkSystem
import com.ore.infinium.util.mapper
import com.ore.infinium.util.require
import com.ore.infinium.util.system

@Wire(failOnNull = false)
class PlayerSystem(private val oreWorld: OreWorld) : IteratingSystem(Aspect.all()) {

    private val mPlayer by require<PlayerComponent>()
    private val mSprite by mapper<SpriteComponent>()

    private val serverNetworkSystem by system<ServerNetworkSystem>()

    private val chunkTimer = OreTimer()

    override fun inserted(entityId: Int) {
        super.inserted(entityId)

        //client does nothing as of yet, with this
        if (oreWorld.isClient()) {
            return
        }

        //initial spawn, send region
        calculateLoadedViewport(entityId)
        sendPlayerBlockRegion(entityId)
    }

    override fun process(entityId: Int) {
        if (oreWorld.isClient()) {
            return
        }

        //clients, for now, do their own collision stuff. mostly.
        //FIXME: clients should simulate their own player's collision with everything and tell the server its
        // position so it can broadcast.
        // but nothing else.
        //server will simulate everything else(except players), and broadcast positions

        //should never ever, ever happen.
        assert(mSprite.has(entityId) && mPlayer.has(entityId))

        val cSprite = mSprite.get(entityId)
        val cPlayer = mPlayer.get(entityId)

        val viewportRect = cPlayer.loadedViewport.rect
        val x = cSprite.sprite.x
        val y = cSprite.sprite.y

        //fixme fixme, we'll fix this when we get to chunking and come up with a proper solution
        if (chunkTimer.resetIfSurpassed(600L)) {
            calculateLoadedViewport(entityId)
        }
    }

    private fun calculateLoadedViewport(playerEntity: Int) {
        val cPlayer = mPlayer.get(playerEntity)
        val cSprite = mSprite.get(playerEntity)

        val loadedViewport = cPlayer.loadedViewport

        val center = Vector2(cSprite.sprite.x, cSprite.sprite.y)
        loadedViewport.centerOn(center, oreWorld)

        serverNetworkSystem.sendPlayerLoadedViewportMoved(playerEntity)

        //todo send only partials depending on direction they're traveling(distance from origin).
        sendPlayerBlockRegion(playerEntity)
    }

    fun sendPlayerBlockRegion(playerEntity: Int) {
        val cPlayer = mPlayer.get(playerEntity)
        val loadedViewport = cPlayer.loadedViewport

        val region = loadedViewport.blockRegionInViewport()

        serverNetworkSystem.sendPlayerBlockRegion(playerEntity, region.x, region.width, region.y,
                                                  region.height)
    }
}
